### DISCUSSION TITLE
Feature request: Display the exact date/time of each action in the list of worflow runs

### DISCUSSION THREAD

**Discussion by oksana-c** [community/community#58663] [Actions] [answered]
### Select Topic Area

Product Feedback

### Body

The exact date/time of GitHub Action execution is shown on hover only. 
The task of visually comparing timestamps of multiple actions is challenging when forensics must be performed quickly.

---
**As a GitHub Actions user, I need to see the exact timestamp of when the action was run at a glance without having to hover over the relative time text, so that I quickly analyze the information.**

---

<img width="1291" alt="Screen Shot 2023-06-21 at 8 46 50 AM" src="https://github.com/community/community/assets/16477724/d99bfe44-dce9-4ba8-8068-ea31f4415000">

**Comment by rgegriff** [comment]
:+1: The current display is fine most of the time; but we use a LOT of actions, and not being able to see exact timestamps in the UI makes it fairly difficult to do forensics when things go wrong.

**Reply by mmustala** [reply]
Exactly. Right now I would need to know what was deployed to our server right before a certain timestamp.

**Reply by mmustala** [reply]
Hey, found out that when you hover the "3 weeks ago" it shows the accurate timestamp. So I can find the deploy I'm looking for.

**Comment by zacmoss-jaspar** [comment]
This would be GREAT.

**Comment by rjcoupe** [comment]
This is shocking in its absence. The UI happily shows you far less useful, yet presumably similarly sourced information like "4 days ago", but appears designed specifically to not show you an actual date and time.

**Reply by zacmoss-jaspar** [reply]
"4 days ago"

"Yeah but when exactly?"

"Dunno lol"

**Reply by jackofsometrade** [reply]
My thoughts exactly.  It's really important to have precision in tools like this.  The irony is that it requires more work to say "1 month ago" than it does to just provide the exact time.  Why is this being hidden from us?  It is important information.

**Comment by realugbun** [comment]
+1 to this request. I'm an Engineering Manager with about 50 devs at the company. We use Github actions for all our deployments and CI/CD. Having an entire month's worth of actions grouped into "last month" is very hard to work with. It's even worse that when you click on the action it still shows you "last month" instead of the date and time it was run.

**Comment by jan0991** [comment]
Simple chrome extension that solves the issue: https://github.com/jan0991/github-actions-full-datetime

**Reply by leitneratselerity** [reply]
This looks promising. It would be helpful if there were some short instructions on how to use the two files in the linked repo or at least a link to instructions.

**Reply by msorens** [reply]
This extension really works!
<img width="553" alt="image" src="https://github.com/community/community/assets/6817500/0f35b891-9ef8-48d8-8bfb-a8f9693e0579">

I had no idea how to install this extension, either... so I asked ~Skynet~, er, ChatGPT:

To install a Chrome extension using a content.js file and a manifest.json file, you can follow these steps:

1. Open Google Chrome and navigate to `chrome://extensions/`.
2. Enable Developer mode by clicking the toggle switch in the top right corner.
3. Click on the <kbd>Load unpacked</kbd> button that appears.
4. Select the folder that contains your extension files, including the **content.js** and **manifest.json** files.
5. Click "Open" to load the extension into Chrome.

(Step 0, of course, is to download the files from @jan0991 's link above.)

**Reply by msorens** [reply]
Found one glitch that needs fixing, though: the extension frequently uses the wrong AM/PM designator. I posted that as [an issue](https://github.com/jan0991/github-actions-full-datetime/issues/1).

**Reply by BitSteve** [reply]
I've forked and improved it to show on all Github pages
https://github.com/BitSteve/github-full-datetime

**Reply by msorens** [reply]
Just tried it and (a) it works and (b) it fixes the AM/PM issue I noted above (by using 24-hour time 😉 ).
Thank you @BitSteve !! 🎉

**Reply by BitSteve** [reply]
LOL. My pleasure. I decided to use 24-hour time since it is shorter.
If I get a chance in the future, maybe make it configurable.

**Reply by wilblack** [reply]
I just installed this. Works great. Thanks.

**Comment by martinkoch-geniebelt** [comment]
+1 - I was just trying to figure out which release caused issues in production, and having the exact build time to compare against my logs would make my job a lot easier.

**Reply by daniel-lopez-puig** [reply]
If you hover the "x days/months/years ago" then you have exact time :smile:

**Reply by martinkoch-geniebelt** [reply]
Thanks. I see it in the HTML if I inspect the source, but actually not on hover (chrome, 120.0.6099.71)

**Reply by martinkoch-geniebelt** [reply]
Well, actually ... now it appears on hover 🤷 :) - thank you.

**Comment by daniel-lopez-puig** [comment]
HOVER the mouse on the "x days/months/years ago" and you have it

I just put it here because if you google it just looking for the exact time and don't care about the feature and you are like me and read in diagonal .

**Comment by linucks** [comment]
The hover isn't working for me [Chrome Version 120.0.6099.234 (Official Build) (arm64)]. I can see a "datetime" and a "title" attribute that have the information, but nothing happens when I hover and when I activate hover in Developer Tools on any of the elements, nothing happens,

**Comment by amnuts** [comment]
The hover is a little annoying to do when you have a lot actions to go over.

The tag GitHub uses is documented at https://github.com/github/relative-time-element.  So if you have a plugin that allows you to run arbitrary js on a page, then you can do something like:

```js
document.getElementsByTagName('relative-time').forEach(node => {
    node.setAttribute('format', 'datetime');
    node.setAttribute('hour', 'numeric');
    node.setAttribute('minute', 'numeric');
    node.setAttribute('second', 'numeric');
})
```

and it'll change the format of the time to be more full.

But I really, really hope GitHub put in a toggle switch for this in the preferences!

**Reply by ttlns** [reply]
> ```js
> document.getElementsByTagName('relative-time').forEach(node => {
>     node.setAttribute('format', 'datetime');
>     node.setAttribute('hour', 'numeric');
>     node.setAttribute('minute', 'numeric');
>     node.setAttribute('second', 'numeric');
> })
> ```

```
TypeError: document.getElementsByTagName(...).forEach is not a function
```

Refactored..
```js
function show_full_time(){
  let s = document.getElementsByTagName('relative-time');
  for (let node of s){
      node.setAttribute('format', 'datetime');
      node.setAttribute('hour', 'numeric');
      node.setAttribute('minute', 'numeric');
      node.setAttribute('second', 'numeric');
  }
}
show_full_time()
```

**Comment by nghhoese** [comment]
Any update on this this would be really nice to have

**Comment by coactive-jeffrey** [comment]
We need exact timestamps for post-mortems. Please add a mouseover or something that gives us the exact time and timezone. The existing "Manually triggered last week" stuff isn't good enough to help us track and debug errors.

**Comment by larmitage-bjss** [comment]
As others have said, you can see this on hover, however it would be nice to be able to see it quicker and easier for skimming through pages of runs
I don't see why this can't be an option in your personal settings the same as timezone and (not sure github has this but other apps do) whether you want to see times in 24 or 12 hour clock

**Comment by kevin930321** [comment]
https://github.com/community/community/assets/78043580/4d229324-c839-4ef2-b347-f999253d1965

When u use the mobile web mode, it'll show a glimpse when u reload it, really weird design.

**Comment by pfoneill** [comment]
+1 for this feature.

In addition it would be nice to have the current date/time exposed in a top-level variable that can be included in the `run-name:`
e.g.
```
name: nightly build
run-name: Nightly build on ${{ github.date }} at ${{ github.time }}
```

**Comment by msorens** [comment]
**Bubbling this up because it solves the problem (as of today 2024.05.30 !!):** https://github.com/orgs/community/discussions/58663#discussioncomment-9608722

In a nutshell, @BitSteve created a Chrome plugin that adds a full date/time next to the existing relative date/time.

**Comment by 2beens** [comment]
Such a simple thing to do, and it's still missing ... we are engineers here working with software, not betting in a casino, exact timestamps are quite important. "5 days ago" tells very little.

**Comment by OscarGil-Sage** [comment]
+1 to the topic. Also in the containers view, as the hover doesn't work there
![image](https://github.com/community/community/assets/62386573/606c4e23-a398-4555-9abd-3249a962d9b8)

I'm going to give a chance to the extension, but I don't know why an easy thing to develop and deploy, and requested, is not present yet.

**Comment by jonapgar-groupby** [comment]
This is what happens when UX designers thinks they know better than their users.

**Comment by stas-lucky** [comment]
Thread is silent... It is so annoying to see this relevant time. I waste so much time on that. Now even hover doesn't work in actions as described in the comments. Please give just timestamp instead of "3 week ago". Or give somehow both options. This drives me crazy, such a simple thing.

**Comment by lkfortuna** [comment]
Hey there from Actions 👋 Thanks for bringing this to our attention, and apologies for the silence. This is a totally understandable request, right now the date/time display is standardized across GitHub and is standard in PRs, Issues, Discussions, Projects. We do not have the power to change it just for our product at this time.
For the bug in the hover option, I have reported that to our internal design team.

**Reply by jonapgar-groupby** [reply]
Standardization strikes again!

**Reply by owattley-rotageek** [reply]
@lkfortuna  Can you confirm that this issue has been escalated to the team that does have the power to make this change? Or can you confirm that it will not be getting addressed at all?

**Comment by rdhzl** [comment]
This issue makes me miss Jenkins.

**Comment by iway1** [comment]
does on hover even work now? not working for me looking at actions

**Reply by rdhzl** [reply]
Not on Chrome. Does on Firefox I believe.

**Reply by FearlessHyena** [reply]
Yea the hover doesn't work on Chrome anymore. It seems to work on Safari though
Wonder if GitHub will ever fix this issue 😞

**Comment by dtrezubSV** [comment]
+1 to this. Relative time should be an addition to exact time (i.e. `2024-23-11 04:37 - 2 weeks ago`)

**Comment by zoqiao** [comment]
+1.  i like the exact date + time

**Comment by ynorsa** [comment]
This would be super useful indeed, particularly considering that the hover does not work in many cases

**Comment by Shreyamitti** [comment]
Hover is not working for me :( 
I can only see 2 weeks ago

**Comment by PatrickHum-at** [comment]
If you use any injected js tool (like Arc boosts) try this:

```
const run = () => {
  document.querySelectorAll('relative-time').forEach(function(element) {
    element.setAttribute("format", "datetime");
  });
}
window.addEventListener('load', setTimeout(run, 200))
```

**Comment by sophie-warner-annalise-ai** [comment]
Firefox extension which fixes this: https://addons.mozilla.org/en-US/firefox/addon/nudiflorum/
(not mine)

**Comment by yyz999** [comment]
For a product mostly used by engineer, it's showing time in xx ago with unstable granularity. 
I really want to hear some opinions from anyone advocate it. Why xx ago is better than exact datetime .

**Comment by adamshawvipps** [comment]
+1 absolute timestamps

**Comment by kevinhpage** [comment]
I'm curious if there is anyone who finds value in the relative display.  I don't understand why we don't have exact times for *all* the items mentioned by @lkfortuna.  I understand this is a system-wide design decision and would be a big change (potentially?).  But who are the users who want to see a vague human-readable sentence?  Maybe it should just go.  That would be my vote.

**Comment by askprasanth** [comment]
+1
This is definitely valuable for auditing and capturing timestamps. It displays perfectly in the ADO pipeline, which is especially helpful during broader product releases. Really appreciate this being in place!

**Comment by Xaeroxe** [comment]
GitHub please fix this. It's absurd that I have to install a browser extension just to get this basic info. I'm an engineer, I want precision from my tooling.

**Comment by onreal** [comment]
If you want to see the exact datetime an action was run, you can right-click on the Actions page and inspect the source. Look for the `<relative-time>` element that renders the “x days ago” text.

Personally, the relative time works well in most cases, but when I do need the exact timestamp, it's right there in the DOM.

That said, +1 to getting this fixed (assuming it's a bug), since the exact time is already sent to the client but just not displayed.

**Comment by qoomon** [comment]
I'm using the solution from above as a bookmarklet, works like charm.
```js
javascript:(function (){  
    let s = document.getElementsByTagName('relative-time');
    for (let node of s){ 
        node.setAttribute('format', 'datetime');      
        node.setAttribute('hour', 'numeric');      
        node.setAttribute('minute', 'numeric');  
        node.setAttribute('time-zone-name', 'shortOffset');   
}})();
```

**Comment by ttlns** [comment]
My suggestion here: At least make it (especially the rounding) consistent. e.g.
<img width="640" alt="17517078520246691483710972350905" src="https://github.com/user-attachments/assets/132b3af3-44dd-4f58-a0c6-0a9ac1ce13b1" />
![17517079591814816617561896916186](https://github.com/user-attachments/assets/3c96785c-b663-462c-837e-c739213fb880)


Chrome extension: [kaliiiiiiiiii/github-accurate-timedelta](https://github.com/kaliiiiiiiiii/github-accurate-timedelta) ([Chrome webstore](https://chromewebstore.google.com/detail/github-accurate-time-disp/fhoalenimobepcaddmalfppleofpkill))

**Comment by timcosgrove** [comment]
The decision to show relative 'human time' has to have been made by a person who never actually has to manage CI/CD. There is *no* situation in working with CI/CD where 'last week' is a useful piece of information.

If you hover over the relative human time, you get the real time. Github is in fact using a web component called `<relative-time>` to wrap the real date & time.

<img width="513" height="197" alt="image" src="https://github.com/user-attachments/assets/91a1e54c-4c8c-4b20-87e2-59d6ba521ec2" />

I strongly suspect this was a UX decision made by someone who does not actually do this work. In most cases, relative 'human' time is in fact what you would want to see. CI/CD is not one of those cases. 

At a very minimum, please give users the option to choose how the time is displayed.

**Comment by atkinsjb** [comment]
Celebrating 2+ years of users complaining about this issue with no update from Github. 🎉🎂🎉

What do you learn when users are so frustrated with your UX that they write their own browser extensions to fix it???

**Reply by lknhd** [reply]
i am f ing furious with this

**Comment by RomanMindlin** [comment]
Are there any estimates?

**Reply by rdhzl** [reply]
If there were it would be something like "In a Few Months", "Next Year", "Two Years From Now".

**Comment by 4wk-** [comment]
@mikeplem that's not the place we want **and need** timestamps.  
This thread is about the "All workflows" page, as shown is the original post https://github.com/orgs/community/discussions/58663#discussion-5323515 
<img width="2582" height="482" alt="image" src="https://github.com/user-attachments/assets/a9d3e193-5f2c-4776-9fe3-39c59a567c25" />

**Reply by mikeplem** [reply]
My mistake.  I am going to delete my comments since it does  not add value to this issue.

**Comment by Quezion** [comment]
Here in the middle of an incident alert, I can't believe GitHub is displaying "last week" on our last service prod run which was _10 days ago._

Please prioritize & fix this. Developers need precision, not to have their hands held with inaccurate human readable timestamps.

**Comment by geoidesic** [comment]
Please prioritise this!

**Comment by eshattow** [comment]
Also see ref.: https://github.com/orgs/community/discussions/41845

**Comment by brainwipe** [comment]
GitHub, I need to know the exact time and date when an action started. This vague nonsense makes it impossible. You've not given us this information after 2 years of complaints? Why not?

**Comment by jbeers** [comment]
Can this just be hotfixed with a tooltip that displays the exact date and time already?

**Comment by Nightblade** [comment]
Looks like they've implemented it.

<img width="213" height="194" alt="image" src="https://github.com/user-attachments/assets/9c32c2b9-214f-4393-b9d6-b92158733c25" />

**Reply by eshattow** [reply]
No, it shows "Last week" and other useless humanisms

**Reply by ttlns** [reply]
For [kaliiiiiiiiii/esdtracker/actions](https://github.com/kaliiiiiiiiii/esdtracker/actions) , I can confirm (matching [discussioncomment-14466846](https://github.com/orgs/community/discussions/58663#discussioncomment-14466846)) the date to be shown in the following format:
1. `Today at 2:55 AM` for recent actions (guess: same day)
2. `Sep 20, 2:47 AM GMT+2`

Important question now is, based on what Timezone it counts as "recent", e.g. same day.

An improvement, yes. But **not a complete fix** 🥲

**Reply by zaataylor** [reply]
👋  Hi there @kaliiiiiiiiii! Regarding:

> Important question now is, based on what Timezone it counts as "recent", e.g. same day.

Here's how it currently works since last Thursday's (September 18th) GA:
- For runs which are _less than an hour old_, we use the existing relative time display.
- For runs _older than an hour, but which occurred on the same day_, we use the `Today at hh:mm` format you listed in your example above.
- For runs _older than that_, i.e. not on the same day, we use the absolute time including the day, month, (year if the run was last year) and timezone.

> An improvement, yes. But not a complete fix 🥲

I agree! All I can say is, we're working on it, and I'm sorry this isn't ideal right now! 😄 We chose the 1 hour threshold and "semi-relative", `Today at hh:mm` time display initially because seeing the exact time as soon as the run started felt a bit awkward from a UX perspective. Ideally, this would probably take the form of something like a toggle-able user setting, _but_ it's not there yet, _and_ we're aware of that.

There are a group of us advocating internally for prioritizing and fixing these kinds of papercuts/product gaps that users like those in this discussion thread run into. I implemented these changes along with a colleague based on this discussion, as I had time, aside from my normal, day-to-day tasks. It gives me immense joy to make small improvements that make things work better for our users, but these changes unfortunately don't happen as fast as you (or I 😄) would like! We welcome constructive feedback on how we can make this better, and we have further ideas/discussion for using this or a similar display format in other high-trafficked UIs.

**Comment by zaataylor** [comment, marked as answer]
👋 😄 

Last Thursday (September 18th), we released a change to improve how the date/time is being represented in Actions workflow executions.

Previously, the exact date/time of a GitHub Action execution was only visible on hover as you all have described in this thread (see the screenshot below, which was pulled from the original discussion comment [here](https://github.com/orgs/community/discussions/58663?sort=new#discussion-5323515)), making it harder to quickly reference when a run occurred.

<img width="2582" height="482" alt="image" src="https://github.com/user-attachments/assets/2e9d4291-850e-4073-9137-2b4b60415853" />

Now, we've introduced a **1 hour** threshold before transitioning from relative to absolute time to improve clarity and efficiency for anyone tracking workflow runs. To strike a good balance between providing enough information for a user to correlate changes in their system over the course of the same day with an exact `hh:mm` time, we did the following:

- For the first hour, you'll still see the relative time e.g. `30 minutes ago`
- After that hour has elapsed, you'll see a semi-relative time e.g. `Today at 8:33 AM`
- Finally, for runs older than the current day, you'll see exact time e.g. `Sep 16, 10:29 AM EDT`

Feel free to leave a comment in this thread with any thoughts or feedback! I worked on this with a colleague alongside my normal, day-to-day work, and we _do_ have ideas for how to improve this, and we welcome continued, _constructive_ feedback. _I know_ the preexisting time display behavior is annoying and makes the experiences of tracking workflow runs unideal (to say the least). We are working to improve it!

**Reply by kevinhpage** [reply]
Thanks so much for your work on this!

**Reply by eshattow** [reply]
> * For the first hour, you'll still see the relative time e.g. `30 minutes ago`

Does the pop-over show the exact time?

> * After that hour has elapsed, you'll see a semi-relative time e.g. `Today at 8:33 AM`

Does the pop-over show the relative time?

> * Finally, for runs older than the current day, you'll see exact time e.g. `Sep 16, 10:29 AM EDT`

Suggest to tri-state relative mode somehow, and an actionable UX feature for selecting a different relative time reference. Weak on the details how this could be implemented in UX but it would make the relative time a useful functional feature. Often when we are saying this relative time feature is annoying what we mean is that it is an obstacle when what we are likely trying to do is solve a problem that uses a different time reference than "now".

Example: Single-click a UI element (for example any of the calendar glyphs but not the text) anywhere on the viewed page to tri-state this for the whole page being viewed:

- Blended
- Relative
- Exact

Search method from UI (clicking calendar icon is perhaps better suited to this; caveat):
- Relative time relative reference
  - Relative to "now"
  - Relative to calendar date selection (defaults to "now" or selected commit context and shows additional reference text "+/-" and the selected date or commit id)
  - Relative to commit

Future preferences:

- Blended time demarcation select how much time elapsed between display of relative and exact time
  - numeric combo box or slider UI input 1-9999
  - unit selection
    - minutes
    - hours
    - days
    - months

**Reply by Nightblade** [reply]
Is there a public-facing change-log for this kind of thing?
